File: ss/OS/SPOS_java/Scheduling/Round_robin.java
import java.util.LinkedList;
import java.util.Queue;

class Process {
    int pid;
    int burstTime;
    
    public Process(int pid, int burstTime) {
        this.pid = pid;
        this.burstTime = burstTime;
    }
}

public class RoundRobinScheduling {
    public static void main(String[] args) {
        Queue<Process> queue = new LinkedList<>();
        
        Process[] processes = {
            new Process(1, 10),
            new Process(2, 5),
            new Process(3, 8),
            new Process(4, 12)
        };

        int quantum = 2; // Time quantum

        for (Process process : processes) {
            queue.add(process);
        }

        System.out.println("Process\tRemaining Time");

        while (!queue.isEmpty()) {
            Process currentProcess = queue.poll();
            int remainingTime = currentProcess.burstTime > quantum ? currentProcess.burstTime - quantum : 0;

            System.out.println(currentProcess.pid + "\t\t" + remainingTime);

            currentProcess.burstTime = remainingTime;

            if (remainingTime > 0) {
                queue.add(currentProcess);
            }
        }
    }
}

File: ss/OS/Scheduling/round_robin.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

struct Process {
    int id;
    int burstTime;
    int remainingTime;
};

int main() {
    int n, timeQuantum;
    cout << "Enter the number of processes: ";
    cin >> n;
    
    cout << "Enter the time quantum: ";
    cin >> timeQuantum;

    vector<Process> processes(n);
    for (int i = 0; i < n; i++) {
        processes[i].id = i + 1;
        cout << "Enter burst time for Process " << i + 1 << ": ";
        cin >> processes[i].burstTime;
        processes[i].remainingTime = processes[i].burstTime;
    }

    queue<Process> readyQueue;
    int currentTime = 0;
    vector<int> turnaroundTime(n);
    vector<int> waitingTime(n);

    while (!readyQueue.empty() || currentTime < n) {
        for (int i = 0; i < n; i++) {
            if (processes[i].remainingTime > 0 && processes[i].burstTime > timeQuantum) {
                processes[i].remainingTime -= timeQuantum;
                currentTime += timeQuantum;
                readyQueue.push(processes[i]);
            } else if (processes[i].remainingTime > 0) {
                currentTime += processes[i].remainingTime;
                processes[i].remainingTime = 0;
                turnaroundTime[i] = currentTime;
            }
        }

        while (!readyQueue.empty() && processes[currentTime % n].remainingTime == 0) {
            readyQueue.pop();
        }
        
        if (!readyQueue.empty()) {
            readyQueue.push(processes[currentTime % n]);
        }
    }

    // Calculate waiting time
    for (int i = 0; i < n; i++) {
        waitingTime[i] = turnaroundTime[i] - processes[i].burstTime;
    }

    // Calculate average waiting time and turnaround time
    double totalWaitingTime = 0;
    double totalTurnaroundTime = 0;
    for (int i = 0; i < n; i++) {
        totalWaitingTime += waitingTime[i];
        totalTurnaroundTime += turnaroundTime[i];
    }
    double averageWaitingTime = totalWaitingTime / n;
    double averageTurnaroundTime = totalTurnaroundTime / n;

    // Display results
    cout << "Process\tBurst Time\tWaiting Time\tTurnaround Time" << endl;
    for (int i = 0; i < n; i++) {
        cout << "P" << processes[i].id << "\t" << processes[i].burstTime << "\t\t" << waitingTime[i] << "\t\t" << turnaroundTime[i] << endl;
    }

    cout << "Average Waiting Time: " << averageWaitingTime << endl;
    cout << "Average Turnaround Time: " << averageTurnaroundTime << endl;

    return 0;
}
